modules/partition-replicator/src/integrationTest/java/org/apache/ignite/internal/partition/replicator/ItColocationTxRecoveryTest.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.ignite.tx.Transaction;
+import org.junit.jupiter.api.Test;
+
+class ItColocationTxRecoveryTest extends ItAbstractColocationTest {
+    private static final long KEY = 1;
+
+    /**
+     * Tests that tx recovery works. Scenario:
+     *
+     * <ol>
+     *     <li>A transaction tx1 is started, it takes a shared lock on a key and never gets finished</li>
+     *     <li>Its coordinator (different from the node hosting the touched partition primary) is stopped, so the transaction becomes
+     *     abandoned</li>
+     *     <li>Transaction tx2 tries to write to the same key, founds an incompatible lock, realizes that it's held by an abandoned
+     *     transaction, and does tx recovery to remove the lock on the partition primary</li>
+     *     <li>tx2 should succeed</li>
+     * </ol>
+     */
+    @Test
+    void abandonedTransactionGetsAbortedOnTouch() throws Exception {
+        assertThat(txConfiguration.abandonedCheckTs().update(600_000L), willCompleteSuccessfully());
+
+        startCluster(3);
+
+        Node node0 = getNode(0);
+
+        // Create a zone with a single partition on every node.
+        int zoneId = createZone(node0, TEST_ZONE_NAME, 1, cluster.size());
+
+        createTable(node0, TEST_ZONE_NAME, TEST_TABLE_NAME1);
+
+        cluster.forEach(Node::waitForMetadataCompletenessAtNow);
+
+        putInitialValue(node0);
+
+        ReplicaMeta primaryReplica = getPrimaryReplica(zoneId);
+
+        Node coordinatorNodeToBeStopped = findAnyOtherNode(primaryReplica);
+        Transaction txToBeAbandoned = coordinatorNodeToBeStopped.transactions().begin();
+        // Trigger a shared lock to be taken on the key.
+        coordinatorNodeToBeStopped.tableManager.table(TEST_TABLE_NAME1)
+                .keyValueView(Long.class, Integer.class)
+                .get(txToBeAbandoned, KEY);
+
+        coordinatorNodeToBeStopped.stop();
+        cluster.remove(coordinatorNodeToBeStopped);
+
+        Node runningNode = cluster.get(0);
+
+        KeyValueView<Long, Integer> kvView = runningNode.tableManager.table(TEST_TABLE_NAME1).keyValueView(Long.class, Integer.class);
+
+        Transaction conflictingTx = runningNode.transactions().begin();
+        assertDoesNotThrow(() -> kvView.put(conflictingTx, KEY, 111));
+    }
+
+    private static void putInitialValue(Node node0) {
+        node0.tableManager.table(TEST_TABLE_NAME1).keyValueView(Long.class, Integer.class).put(null, KEY, 42);
+    }
+
+    private ReplicaMeta getPrimaryReplica(int zoneId) {
+        Node node = cluster.get(0);
+
+        CompletableFuture<ReplicaMeta> primaryReplicaFuture = node.placementDriverManager.placementDriver().getPrimaryReplica(
+                new ZonePartitionId(zoneId, 0),
+                node.hybridClock.now()
+        );
+
+        assertThat(primaryReplicaFuture, willCompleteSuccessfully());
+        return primaryReplicaFuture.join();
+    }
+
+    private Node findAnyOtherNode(ReplicaMeta primaryReplica) {
```

Comment:
`findAnyNonPrimaryColocatedNode` or `findAnyNonPrimaryNode` is better, but opinionated.

Reply:
The line where it's used looks like this:

```
Node coordinatorNodeToBeStopped = findAnyOtherNode(primaryReplica);
```

So the reader of the test scenario sees that we are looking for any node that is not the primary replica; it seems that this doesn't need to be made even more obvious :)